File: 6. Binary Search/33. Search in Rotated Sorted Array/Solution.cpp
#include <vector>

class Solution {
public:
    int search(std::vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right)
        {
            int index = left + (right - left) / 2;
            int middle = nums[index];

            if (middle == target) return index;

            if (middle <= nums[right])
            {
                if (middle < target && target <= nums[right]) left = index + 1;
                else right = index - 1;
            }
            else
            {
                if (nums[left] <= target && target < middle) right = index - 1;
                else left = index + 1;
            }
        }

        return -1;
    }
};

File: 6. Binary Search/33. Search in Rotated Sorted Array/Solution.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while (left <= right):
            middle = left + (right - left) // 2
            middle_num = nums[middle]

            if middle_num == target: return middle

            if (nums[left] <= middle_num):
                if (nums[left] <= target < middle_num):
                    right = middle - 1
                else:
                    left = middle + 1
            else:
                if (middle_num < target <= nums[right]):
                    left = middle + 1
                else:
                    right = middle - 1
        
        return -1
        